staging/src/k8s.io/apiserver/pkg/server/filters/timeout.go

```diff
@@ -146,24 +148,26 @@ type timeoutWriter interface {
 	timeout(*apierrors.StatusError)
 }
 
-func newTimeoutWriter(w http.ResponseWriter) timeoutWriter {
+func newTimeoutWriter(w http.ResponseWriter) (timeoutWriter, http.ResponseWriter) {
 	base := &baseTimeoutWriter{w: w}
-
-	_, notifiable := w.(http.CloseNotifier)
-	_, hijackable := w.(http.Hijacker)
-
-	switch {
-	case notifiable && hijackable:
-		return &closeHijackTimeoutWriter{base}
-	case notifiable:
-		return &closeTimeoutWriter{base}
-	case hijackable:
-		return &hijackTimeoutWriter{base}
-	default:
-		return base
-	}
+	wrapped := responsewriter.Wrap(base, responsewriter.Callback{
+		CloseNotify: func() <-chan bool {
+			return base.closeNotify()
+		},
+		Flush: func() {
+			base.flush()
+		},
+		Hijack: func() (net.Conn, *bufio.ReadWriter, error) {
+			return base.hijack()
+		},
```

Comment:
This is an unpleasant amount of text just to say to use the implementations from `base`.  A shorter way to say it would be as follows.

```
    CloseNotify: base.closeNotify,
    Flush: base.flush,
    Hijack: base.hijack
```

We could avoid even having to say that if `responsewriter.Wrap` itself would use those methods if the given decorator has them.

Reply:
I am thinking of the following
```
type WrappedFlusher interface {
	WrappedFlush()
}
```

the decorator can optionally implement the `WrappedFlusher` interface and we will call it as follows

```
func (wr *withCloseNotifyAndFlush) Flush() {
	if flusher, ok := wr.ResponseWriterDecorator.(WrappedFlusher); ok {
		flusher.WrappedFlush()
		return
	}

	inner := wr.ResponseWriterDecorator.Unwrap()
	inner.(http.Flusher).Flush()
}
```

this completely eliminates the need for `Callback` 
```
	base := &baseTimeoutWriter{w: w}
	wrapped := responsewriter.Wrap(base)
```

@MikeSpreitzer thoughts?

Reply:
Your answer about "Flush" included "Hijack" at one point; I assume that is a typo.

We do not even need `WrappedFlusher`, we can just use `http.Flusher`.  The key idea is for the responsewriter wrapper to dispatch to the middle object if it implements that interface.

A remaining irritant is that dynamic type tests are not totally cheap.  That could be eliminated by shifting all the dynamic type testing left to the construction, using a wrapper type that has exactly the right methods.  That would make for more verbose construction and involve more types.

Reply:
> Your answer about "Flush" included "Hijack" at one point; I assume that is a typo.

yes, it was a typo.

> We do not even need WrappedFlusher, we can just use http.Flusher. The key idea is for the responsewriter wrapper to dispatch to the middle object if it implements that interface.

It's true for `http.Flusher` and `http.CloseNotifier`,  on the contrary http2 never implements `http.Hijacker`. So a decorator should not implement `http.Hijacker`, if it does then it is erroneously extending the original `ResponseWriter` object provided by net/http2.

I think we can simplify with:
- a decorator can always implement `http.CloseNotifier` and `http.Flusher` (since both these interfaces are implemented by http1 and http2)
- a decorator should not implement `http.Hijacker` since it is never implemented by http2, instead a decorator should implement `WrappedHijacker` interface if it wants to tap into `Hijack`.

- the wrapper will check for `http.CloseNotifier`, `http.Flusher` and a `WrappedHijacker`

Reply:
I think some terminology clarification may help here; it is confusing because a layer of decoration is coming from this package and another layer is coming from this package's client.  So let us speak of:
- the "inner" ResponseWriter, which might come from an HTTP server
- the "middle" ResponseWriter, which comes from the client of this package and you have been calling "the decorator"
- the "outer" ResponseWriter, which is created when this package wraps one or both of its structs around the middle ResponseWriter.

Now, first question: do we require the inner ResponseWriter to implement `http.CloseNotifier` and `http.Flusher`?
- if so then let's define an interface that is the combination of `http.ResponseWriter`, `http.CloseNotifier`, and `http.Flusher` and require the inner rw to implement that interface.
- if not then we have the same problem with CloseNotify and Flush as we have with Hijack.

It is OK if the middle ResponseWriter implements `http.Hijack` even when the inner ResponseWriter does not, because the wrapping done by this package depends on what the inner rw implements.